src/shims/x86/mod.rs

```diff
@@ -1133,6 +1140,58 @@ fn pmulhrsw<'tcx>(
     Ok(())
 }
 
+/// Perform a carry-less multiplication of two 64-bit integers, selected from a and b according to imm8, and store the results in dst.
+///
+/// <https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm_clmulepi64_si128>
+fn pclmulqdq<'tcx>(
+    this: &mut crate::MiriInterpCx<'tcx>,
+    left: &OpTy<'tcx>,
+    right: &OpTy<'tcx>,
+    imm8: &OpTy<'tcx>,
+    dest: &MPlaceTy<'tcx>,
+) -> InterpResult<'tcx, ()> {
+    assert_eq!(left.layout, right.layout);
+    assert_eq!(left.layout.size, dest.layout.size);
+
+    // Transmute to `[u64; 2]`
+
+    let array_layout = this.layout_of(Ty::new_array(this.tcx.tcx, this.tcx.types.u64, 2))?;
+    let left = left.transmute(array_layout, this)?;
+    let right = right.transmute(array_layout, this)?;
+    let dest = dest.transmute(array_layout, this)?;
+
+    let imm8 = this.read_scalar(imm8)?.to_u8()?;
+
+    let left_low = this.read_scalar(&this.project_index(&left, 0)?)?.to_u64()?;
+    let left_high = this.read_scalar(&this.project_index(&left, 1)?)?.to_u64()?;
+
+    let right_low = this.read_scalar(&this.project_index(&right, 0)?)?.to_u64()?;
+    let right_high = this.read_scalar(&this.project_index(&right, 1)?)?.to_u64()?;
+
+    let temp1 = if (imm8 & 0x01) == 0 { left_low } else { left_high };
+    let temp2 = if (imm8 & 0x10) == 0 { right_low } else { right_high };
+
+    // Perform carry-less multiplication
```

Comment:
What exactly is "carry-less multiplication", and why does it involve XOR...?

I've read the linked docs but that looks even more complicated and not even obviously equivalent -- there are 2 loops here, and a funny `TEMP1[0] and TEMP2[i]` initialization.

Reply:
That's a good reference, can you add that to the code?